antenna/util.py

```diff
@@ -293,33 +288,36 @@ def _retry_fun(*args, **kwargs):
 
                     # The return value is "bad", so we log something and then
                     # do another iteration.
-                    if module_logger is not None:
-                        module_logger.warning(
-                            "%s: bad return, retry attempt %s",
-                            fun.__qualname__,
-                            attempts,
+                    log_warning(
+                        "%s: bad return, retry attempt %s", fun.__qualname__, attempts,
+                    )
+
+                    # If last attempt,
+                    if not wait_times:
+                        raise MaxAttemptsError(
+                            "Maximum retry attempts; last return %r." % ret, ret
                         )
 
-                    # If last attempt, then raise MaxAttemptsError
-                    if next_wait is None:
-                        raise MaxAttemptsError("Maximum retry attempts.", ret)
-
                 except retryable_exceptions as exc:
-                    # Retryable exception is thrown, so we log something and
-                    # then do another iteration.
-                    if module_logger is not None:
-                        module_logger.warning(
-                            "%s: exception %s, retry attempt %s",
-                            fun.__qualname__,
-                            exc,
-                            attempts,
-                        )
-
-                    # If last attempt, re-raise the exception thrown
-                    if next_wait is None:
+                    # If it's a MaxAttemptsError, re-raise that
+                    if isinstance(exc, MaxAttemptsError):
```

Comment:
Funny story. So if line 305 in the original code raises a `MaxAttemptsError` and the caller didn't specify which exceptions to retry on, then it would retry on `MaxAttemptsError` from a retryable return. That's definitely not what we want.